### ISSUE TITLE
API considerations

### ISSUE THREAD

**Issue by sastan** [tw-in-js/twind#1]
## Taxomony

- `token`: things that can be passed to `tw` (string, array, object, falsey values)
- `rules`: `rule`, `group`
- `rule`: `variants`, `directive`
- `directive`: `negate`, `plugin`, `params`
- `negate`: should a theme value be negated
- `plugin`: `utility`, `component`
- `params`: parameters for plugins; usually the key in the theme

## Architecture

0. setup - create theme and stuff

**Phases** (details below)

1. **parse** rules
2. **translate** each rule to css object using plugins
3. **decorate** add variant rules to the css object (basicly wrapping it again and again)
4. **serialize** convert css object into css declarations with precedence and hash
5. **inject** css rules

Feature evaluation based on:

- size limit (current stats: 10.5kb)
- performance budget

## API

_common types used below_
```js
type Falsy = '' | 0 | -0 | false | null | undefined

interface TokenGrouping extends Record<string, Token> {}

type Token = string | TokenGrouping | Token[] | Falsy

interface Context {
  /** allow composition */
  tw: TW
  
  /** access to theme values */
  theme(section: string): ThemeSection
  theme(section: string, key: string | string[]): string | undefined
  theme(section: string, key: string | string[], defaultValue: string): string

  /** create unique identifier (group, custom properties) */
  tag(key: string): string
}
```

### tw

> converts rules into CSS class names

- exposed  named export
- supports
  - tagged template where replacements can be interpolations can be strings, array, objects, falsey
  - variadic strings, array, objects, falsey

```js
interface TW {
  (strings: string[], ...replacements: Token[]): string
  (...rules: Token[]): string
}
```

### setup

> customize `tw`

- default to dev settings
    - with example for recommended prod settings
- setup can be called only once
  - warn setup called after tw

```js
interface Setup {
  preflight?: boolean | Preflight
  
  mode?: Mode
  
  theme?: Partial<Theme> | ((defaultTheme: ThemeResolver) => Partial<Theme>)

  hash?: boolean | Hasher
  
  plugins?: Record<string, string | CSSRules | Plugin>
  
  injector?: Injector
  
  prefix?: boolean | Prefix
}
```

#### preflight

> Add global styles before first rule is injected
> **default**: tailwind preflight

- `false`
- function - allow to use a different reset

```js
interface Preflight {
  (preflight: CSSRules, context: Context): CSSRules | undefined
}
```

#### mode

> customize behaviour
> **default**: warn

- warn
- strict
- play

```js
interface Mode {
  /**
   * Notify error (missing plugin, duplicate directives? )
   *
   * Why id?
   * - can generate an url with more info
   * - reduce bundle size by omitting large error messages
   */
  report(info: {id: string}, context: Context): void
  
  /** Called for unknown theme values */
  unknown(
    section: string,
    key: string,
    optional: boolean,
    context: Context,
  ): string | undefined
}
```

#### hash

> **default**: `false`

- boolean

#### theme

> **default**: tailwind

- mergeing and extending like tailwind compatible

#### plugins

> allows to define/override plugins
> **default**: core plugins

Plugins are searched for by name using the longest prefix before a dash (`"-"'`). The name and the remaining parts (splitted by a dash) are provided as first argument to the plugin function. For example if the directive is `bg-gradient-to-t` the following order applies:

| Plugin             | Params                           |
| ------------------ | ------------------------------- |
| `bg-gradient-to-t` | `[]`          |
| `bg-gradient-to`   | `["t"]`       |
| `bg-gradient`      | `["to", "t"]`    |
| `bg`               | `["gradient", "to", "t"]` |

**Arguments**

1. params
2. context
3. id

**Return Types**
- CSSRules: css rules to inject (must use `&` for nested selectors)
- string: class name to use
  - `group` and `lead` return a marker class name that should be used as is
- falsy: no result -> `mode.unknown()` then `mode.report()`

```js
interface Plugin {
  /**
   * Creates CSSRules based on `parts`
   */
  (params: string[], context: Context, id: string): CSSRules | string | Falsy
}
```

**Examples**

```js
const plugins = {
  block: { display: 'block' },

  placeholder: (params, { theme }) => {
    const value = theme('placeholderColor', params)
  
    return value && {
      '&::placholder': {
        color: value
      },
    }
  },

  rotate: (params, { theme }) => {
    const value = theme('rotate', tail(params))
  
    return value && {
      '--tw-rotate': value,
      transform: [
        `rotate(${value})`,
        `translateX(var(--tw-translate-x),0)) translateY(var(--tw-translate-y),0) rotate(var(--tw-rotate,0))`
      ],
    }
  },
  
  /** use tag to create marker class */
  group: (params, { tag }) => tag('group'),
  
  /** use tw to compose */
  card: (params, { tw }) => tw`max-w-md mx-auto bg-white`
}
```

#### injector

> insert CSS rules into runtime
> **default**: based on enviroment

- dom
- collect - for testing in @tw-in-js/test
- ssr - move to @tw-in-js/server

```js
interface Injector {
  insert(rule: string, index: number): void
  delete(index: number): void
}
```

#### prefix

> CSS prefixing 
> **default**: based on tiny-css-prefixer

- `false`
- function

```js
interface Prefixer {
  (property: string, value: string): string
}
```

## Browser Support

- IE11: section how to make it work
- document used modern APIs which need to be polyfilled
  - Math.imul
- reset as extra package
- fallbacks included with deprection notice

### Parse

> parse tokens in rules

```js
interface Parse {
  (strings: string[], replacements: Token[]): Rule[]
  (...rules: Token[]): Rule[]
}

type Token = string | Falsy | Record<string, Token> | Token[]

interface Rule {
  /** ["sm", "dark", "hover"] */
  variants: string[]
  
  /** "text-sm", "-rotate-45" */
  directive: string
}
```

```ebnf
character = letter | digit;
identifier = character, { character };
whitespace = ' ' | '\t' | [ '\r' ], '\n';

(*
- "h-full bg-purple-500 rotate-3 scale-95"
- "w(1/2 sm:1/3 lg:1/6) p-2"
- "divide(y-2 blue-500 opacity(75 md:50))"
- "ring(& ping-700 offset(4 ping-200))"
- "rotate(-3 hover:6 md:(3 hover:-6))"
*)
rules =
    { whitespace },
    group | rule,
    { { whitespace }-, group | rule },
    { whitespace } ;

group = variantGroup | directiveGroup;

variantGroup =
    { variant, ":" }-,
    '(',
        { rules }-,
    ')';

directiveGroup =
    { variant, ":" },
    plugin,
    '(',
        { "&" | rules }-,
    ')';

rule = { variant, ":" }, directive;

variant = identifier;

directive = [ "-" ], plugin, { "-", params };

plugin = identifier, { "-", identifier };

params = param, { "-", param };

param = identifier | "_" | "/" | ".";
```

### Translate

> translate rules into CSSRules using plugins

- no on-colors - maybe as plugin

**Why CSSRules?**

- can be typed with [csstype](https://www.npmjs.com/package/csstype)
- allow to implement advanced directives like `container`, `placeholder`, `divide`, `prose`, ...
- allows to add global default custom properties - this may lead to allow automatic integration tests comparing our result with the ones from tailwind

```js
interface Translate {
  (rule: Rule, context: Context): CSSRules
}

import * as CSS from 'csstype'

interface CSSProperties extends CSS.PropertiesFallback, CSS.PropertiesHyphenFallback {}

interface CSSRules extends CSSProperties {
  /** @media, @supports, @keyframes, ... */
  [`@${string}`]: CSSRules
  
  /** pseudo class
   * maybe -> could be implement using '&:'
   * watch out for ':root' -> that could use '*' instead
   */
  [`:${string}`]: CSSRules

  /** maybe allow '&' everywhere */
  [`${string}&${string}`]: CSSRules

  /** global defaults */
  '*': CSSProperties
}
```

### Inject

> insert CSSRules into the enviroment returning a string with class names

- not atomic
- position based on precedence
- hash class names and CSS custom properties

```js
interface Inject {
  /** we need the parsed rule for class name generation and precedence calculation */
  (css: CSSRules, rule: Rule | undefined, context: Context): string
}
```

## Packages

> monorepo

name: why not go with ocean/beamwind

- @tw-in-js/core - not really happy with the name; but i have no alternative
- @tw-in-js/prose
- @tw-in-js/server - for ssr
- @tw-in-js/typescript-plugin

## Development

- typescript
- es2020 syntax, transpile with esbuild
- snowpack preview

## Tailwind Differences

- `group-*` variant works for **every** pseudo class 
   This allows to use `group-focus-with` or `group-active`
- automatically infers negated values - they do _not_ need to be in the theme config
- `bg-gradient-to-*` is built-in, no need to configure these
- `text-underline`, `text-no-underline`, `text-line-through`, `text-uppercase`, `text-lowercase` and `text-capitalize`: this allows grouping of text directives like `text(lg red-500 capitalize underline)`
- `font-italic` and `font-no-italic`: this allows grouping of font directives like `font(sans italic bold)`
- `border` and `divide` allow to combine positions (`t`op, `r`righ, `l`eft, `b`ottom)

  - `tr` - `top` & `right`
  - `brl` - `bottom`, `right` and `left`

  > **Note** `x` and `y` can not be combined.

- `rotate`, `scale` , `skew` and `translate` provide a fallback for IE 11

  > `transform rotate-45` works but when using `transform rotate-45 scale-150` only one of both is applied.

### IE 11 compatibility

> Some new tailwind features use [CSS Variables (Custom Properties)](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties) and are therefore not compatible with [IE 11](https://caniuse.com/css-variables).

tw-in-js includes fallbacks for the following directives which mimic [Tailwind v1](https://v1.tailwindcss.com/) behavior:

- Color Opacity
  - [border-opacity-\*](https://tailwindcss.com/docs/border-opacity)
  - [bg-opacity-\*](https://tailwindcss.com/docs/background-opacity)
  - [text-opacity-\*](https://tailwindcss.com/docs/text-opacity)
  - [placeholder-opacity-\*](https://tailwindcss.com/docs/placeholder-opacity)
- Reversing Children Order
  - [divide-\*-reverse](https://tailwindcss.com/docs/divide-width#reversing-children-order)
  - [space-\*-reverse](https://tailwindcss.com/docs/space#reversing-children-order)
- `rotate`, `scale` , `skew` and `translate` can only be used alone

  > `rotate-45` works but when using `rotate-45 scale-150` only one of both is applied. In that case you must use `transform`: `transform rotate-45 scale-150`

Some directive only work with CSS Variables and are not supported in IE 11:

- [Ring](https://tailwindcss.com/docs/ring-width)

## Ideas

- react to manual class name change in dev-tools and generate that directive
- live theme updates
- track used theme values and re-translate
- @tw-in-js/styled - like styled components
- @tw-in-js/eslint-plugin - nice to have
- @tw-in-js/tw.macro - maybe?
- svelte, preact, react, vue demos
